src/utils/upload.rs

```diff
@@ -0,0 +1,445 @@
+        pb.enable_steady_tick(100);
+        pb.set_style(
+            ProgressStyle::default_spinner()
+                .tick_chars("/|\\- ")
+                .template(
+                    "{spinner} Looking for symbols... {msg:.dim}\
+                     \n  symbol files found: {prefix:.yellow}",
+                ),
+        );
+
+        let mut batch = vec![];
+        while !self.is_filled(&batch) && !self.found_all() {
+            match self.next_object(&pb)? {
+                Some(sym) => {
+                    batch.push(sym);
+                    pb.set_prefix(&format!("{}", batch.len()));
+                }
+                None => break,
+            }
+        }
+
+        pb.finish_and_clear();
+        Ok(if batch.len() == 0 { None } else { Some(batch) })
+    }
+}
+
+impl<'a> Iterator for BatchedObjectWalker<'a> {
+    type Item = Result<ObjectBatch>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        invert_result(self.next_batch())
+    }
+}
+
+/// Options for uploading debug symbols to Sentry.
+pub struct UploadOptions {
+    api: Api,
+    org: String,
+    project: String,
+    max_size: u64,
+}
+
+impl UploadOptions {
+    /// Creates an `UploadOptions` from command line arguments.
+    pub fn from_cli(matches: &ArgMatches) -> Result<UploadOptions> {
+        let config = Config::get_current();
+        let (org, project) = config.get_org_and_project(matches)?;
+        let max_size = config.get_max_dsym_upload_size()?;
+
+        Ok(UploadOptions {
+            api: Api::new(),
+            org: org,
+            project: project,
+            max_size,
+        })
+    }
+
+    /// Returns the `Api` instance used to upload debug symbols.
+    pub fn api(&self) -> &Api {
+        &self.api
+    }
+
+    /// Returns the Sentry organization slug.
+    pub fn org(&self) -> &str {
+        &self.org
+    }
+
+    /// Returns the Sentry project slug.
+    pub fn project(&self) -> &str {
+        &self.project
+    }
+
+    /// Returns the maximum size for symbol uploads to Sentry.
+    pub fn max_size(&self) -> u64 {
+        self.max_size
+    }
+}
+
+impl fmt::Debug for UploadOptions {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("UploadOptions")
+            .field("org", &self.org)
+            .field("project", &self.project)
+            .field("max_size", &self.max_size)
+            .finish()
+    }
+}
+
+/// Returns an object batch that only includes objects missing on Sentry.
+fn filter_missing_syms(batch: ObjectBatch, context: &UploadOptions) -> Result<ObjectBatch> {
+    info!("Checking for missing debug symbols: {:#?}", &batch);
+
+    let missing_checksums = {
+        let checksums = batch.iter().map(|ref s| s.checksum.as_str()).collect();
+        context.api().find_missing_dsym_checksums(&context.org(), &context.project(), &checksums)?
+    };
+
+    let missing = batch
+        .into_iter()
+        .filter(|sym| missing_checksums.contains(&sym.checksum))
```

Comment:
While rewriting I found that `upload-breakpad` code previously did not deduplicate identical checksums. For Breakpad and ELF, this doesn't matter as each object is in its own file. However, to support Mach-O fat objects, we need to either change this here or change the way objects are Zipped completely.